consensus/cometbft/service/finalize_block.go

```diff
@@ -92,14 +100,65 @@ func (s *Service) finalizeBlockInternal(
 		return nil, err
 	}
 
+	if req.Height == sbtEnableHeight {
+		s.cmtConsensusParams.Feature.SBTEnableHeight = sbtEnableHeight
+	}
+
 	cp := s.cmtConsensusParams.ToProto()
 	return &cmtabci.FinalizeBlockResponse{
 		TxResults:             txResults,
 		ValidatorUpdates:      valUpdates,
 		ConsensusParamUpdates: &cp,
+		NextBlockDelay:        s.nextBlockDelay(req),
 	}, nil
 }
 
+func (s *Service) nextBlockDelay(req *cmtabci.FinalizeBlockRequest) time.Duration {
+	// c0. SBT is not enabled => use the old block delay.
+	if s.cmtConsensusParams.Feature.SBTEnableHeight <= 0 {
+		return constBlockDelay
+	}
+
+	// c1. current height < SBTEnableHeight => wait for the upgrade.
+	if req.Height < s.cmtConsensusParams.Feature.SBTEnableHeight {
+		return constBlockDelay
+	}
+
+	// c2. current height == SBTEnableHeight => initialize the block delay.
+	if req.Height == s.cmtConsensusParams.Feature.SBTEnableHeight {
+		s.blockDelay = &BlockDelay{
+			InitialTime:       req.Time,
+			InitialHeight:     req.Height,
+			PreviousBlockTime: req.Time,
+		}
+		return constBlockDelay
+	}
+
+	// c3. current height > SBTEnableHeight
+	if !(req.Height > s.cmtConsensusParams.Feature.SBTEnableHeight) {
+		panic(fmt.Errorf("Expected height %d > SBTEnableHeight: %d", req.Height, s.cmtConsensusParams.Feature.SBTEnableHeight))
+	}
+
+	// c3.1
+	//
+	// The upgrade was successfully applied and the block delay is set.
+	if s.blockDelay != nil {
+		return s.blockDelay.Next(req.Time, req.Height)
+	} else {
+		// c3.2
+		//
+		// Looks like we've skipped SBTEnableHeight (probably restoring from the
+		// snapshot) => reinitialize the block delay. Treat the current block as a
+		// checkpoint.
+		s.blockDelay = &BlockDelay{
+			InitialTime:       req.Time,
+			InitialHeight:     req.Height,
+			PreviousBlockTime: req.Time,
+		}
+		return constBlockDelay
+	}
```

Comment:
Can this case even happen?
Here`req.Height > s.cmtConsensusParams.Feature.SBTEnableHeight` so any node which has duly processed the activation height has `blockDelay` duly stored. If we are restoring the state, must we not find `blockDelay` duly stored?

Reply:
[`404e20b` (#2422)](https://github.com/berachain/beacon-kit/pull/2422/commits/404e20b4a64c5f7a646d0098a56db8d55fc2d67c)